src/main/java/us/kbase/auth/client/internal/TokenCache.java

```diff
@@ -0,0 +1,150 @@
+	
+	/* should really handle tokens as char[] instead of String so the
+	 * char[] can be wiped when it's no longer needed. But since they're going
+	 * over the wire they're probably already strings or will be converted to
+	 * Strings at some point or another.
+	 */
+	
+	// only non-final so can be tested via reflection.
+	private static int MAX_AGE_MS = 5 * 60 * 1000; // 5 min
+	
+	final private int size;
+	final private int maxsize;
+	final private Map<String, UserDate> cache;
+	
+	/**
+	 * Create a new TokenCache.
+	 * @param size the nominal size of the cache in tokens
+	 * @param maxsize the maximum size of the cache in tokens
+	 */
+	public TokenCache(int size, int maxsize) {
+		if (size < 1 || maxsize < 1) {
+			throw new IllegalArgumentException("size and maxsize must be > 0");
+		}
+		if (size >= maxsize) {
+			throw new IllegalArgumentException("size must be < maxsize");
+		}
+		this.size = size;
+		this.maxsize = maxsize;
+		cache = new ConcurrentHashMap<String, UserDate>(maxsize);
+	}
+	
+	/** Get a token from the cache given its string. Returns null if the 
+	 * cache does not contain the token.
+	 * @param token the token string.
+	 * @return an AuthToken.
+	 */
+	public AuthToken getToken(final String token) {
+		if (token == null || token.isEmpty()) {
+			throw new IllegalArgumentException(
+					"token cannot be null or empty");
+		}
+		final UserDate ud = cache.get(getTokenDigest(token));
+		if (ud == null) {
+			return null;
+		}
+		if (new Date().getTime() - ud.date > MAX_AGE_MS) {
+			return null;
+		}
+		return new AuthToken(token, ud.user);
+	}
+	
+	private String getTokenDigest(final String token) {
+		final MessageDigest digest;
+		try {
+			digest = MessageDigest.getInstance("SHA-256");
+		} catch (NoSuchAlgorithmException e) {
+			throw new RuntimeException("Pretty sure SHA-256 is known, " +
+					"something is very broken here", e);
+		}
+		final byte[] d = digest.digest(token.getBytes(StandardCharsets.UTF_8));
+		final StringBuilder sb = new StringBuilder();
+		for (final byte b : d) {
+			sb.append(String.format("%02x", b));
+		}
+		return sb.toString();
+	}
+	
+	/**
+	 * Add a token to the cache. This method assumes the token is valid.
+	 * @param token the token to add
+	 */
+	public void putValidToken(AuthToken token) {
+		if (token == null) {
+			throw new NullPointerException("token cannot be null");
+		}
+		cache.put(getTokenDigest(token.getToken()),
+				new UserDate(token.getUserName()));
+		synchronized (cache) { // block here otherwise all threads may start sorting
+			if (cache.size() <= maxsize) {
+				return;
+			}
+			List<DateToken> dts = new ArrayList<DateToken>();
+			for (String s: cache.keySet()) {
+				dts.add(new DateToken(cache.get(s).date, s));
+			}
+			Collections.sort(dts);
+			for(int i = size; i < dts.size(); i++) {
+				cache.remove(dts.get(i).token);
+			}
+		}
+	}
+}
+
+class UserDate {
+	String user;
+	long date;
+	
+	UserDate(String user) {
+		super();
```

Comment:
Do we need super() here?

Reply:
Nope, not sure why that's there other than it's decades old code. That being said it's not wrong per se so I'm calling ROE on this one

Reply:
Maybe that class was inherited from a super class. Now we know this is not the case, we should remove it. Otherwise it will be there forever for the same reason.

Reply:
If it stays there forever there will be no negative impact at all though

Reply:
Eh, I just deleted it locally anyway. It'll show up in a later PR

Reply:
> Maybe that class was inherited from a super class.

Technically all classes other than Object inherit from a super class